src/workerd/jsg/modules-new.h

```diff
@@ -0,0 +1,580 @@
+  // evaluation callback to set the exports of the module.
+  class ModuleNamespace final {
+  public:
+    explicit ModuleNamespace(v8::Local<v8::Module> inner,
+                             kj::ArrayPtr<const kj::String> namedExports);
+    KJ_DISALLOW_COPY_AND_MOVE(ModuleNamespace);
+
+    bool set(Lock& js, kj::StringPtr name, JsValue value) const;
+    bool setDefault(Lock& js, JsValue value) const;
+
+  private:
+    v8::Local<v8::Module> inner;
+    kj::HashSet<kj::StringPtr> namedExports;
+  };
+
+  // The EvaluateCallback is used to evaluate a synthetic module. The callback
+  // is called after the module is resolved and instantiated.
+  using EvaluateCallback =
+      kj::Function<bool(Lock&, const Url& specifier,
+                        const ModuleNamespace&,
+                        const CompilationObserver&)>;
+
+  static kj::Own<Module> newSynthetic(
+      Url specifier,
+      Type type,
+      EvaluateCallback callback,
+      kj::Array<kj::String> namedExports = nullptr);
+
+  // Creates a new ESM module that takes ownership of the given code array.
+  // This is generally used to construct ESM modules from a worker bundle.
+  static kj::Own<Module> newEsm(
+      Url specifier,
+      Type type,
+      kj::Array<const char> code,
+      Flags flags = Flags::ESM);
+
+  // Creates a new ESM module that does not take ownership of the given code
+  // array. This is used to construct ESM modules from compiled-in built-in
+  // modules.
+  static kj::Own<Module> newEsm(
+      Url specifier,
+      Type type,
+      kj::ArrayPtr<const char> code);
+
+  // The following methods are used to create the evaluation callbacks for various
+  // kinds of common simple synthetic module types. The module registry is not
+  // limited to just these kinds of modules, however. These are just the most
+  // common.
+
+  static EvaluateCallback newTextModuleHandler(kj::Array<const char> data)
+      KJ_WARN_UNUSED_RESULT;
+  static EvaluateCallback newDataModuleHandler(kj::Array<kj::byte> data)
+      KJ_WARN_UNUSED_RESULT;
+  static EvaluateCallback newJsonModuleHandler(kj::Array<const char> data)
+      KJ_WARN_UNUSED_RESULT;
+  static EvaluateCallback newWasmModuleHandler(kj::Array<kj::byte> data)
+      KJ_WARN_UNUSED_RESULT;
+
+  // An eval function is used for CommonJS style modules (including Node.js compat
+  // modules. The expectation is that this method will be called when the CommonJS
+  // style module is evaluated (e.g. within the EvaluationCallback).
+  static Function<void()> compileEvalFunction(
+      Lock& js,
+      kj::StringPtr code,
+      kj::StringPtr name,
+      kj::Maybe<JsObject> compileExtensions,
+      const CompilationObserver& observer) KJ_WARN_UNUSED_RESULT;
+
+  // Some modules may need to protect against being evaluated recursively. The
+  // EvaluatingScope class makes it possible to guard against that, throwing an
+  // error if there's already an active evaluation happening.
+  class EvaluatingScope {
```

Comment:
nit: We can use more `final` statements here, `EvaluatingScope` and several subclasses come to mind.